### DISCUSSION TITLE
non async interface

### DISCUSSION THREAD

**Discussion by nbraun-wolf** [encode/databases#419] [Q&A] [answered]
Hi, 

I have a situation where some bits of code don't support async. 
Give the fact that it's using SQL alchemy engine under the hood and I also have already sync support by installing psycopg2 to use it with alembic, I was wondering if this package is exposing some kind of non async interface that could be used when there is unfortunately no other way.

**Comment by aminalaee** [comment, marked as answer]
If I understand correctly, you want to use `databases` for sync queries?

As far as I know, that's not possible.

BTW, why do you need that? Sqlalchemy and psycopg2 should be enough for that case.

**Reply by nbraun-wolf** [reply]
I have the problem that I actually want to use async code, but there is one library that doesn't have an anyc interface, https://pypi.org/project/paho-mqtt/. I am looking for ways to be able to interact with the database in its message callbacks.

For the rest of the time I would be able to use async code, since I am using fastapi. But those message callbacks bring me headache. Alternativly, I have tried scheduling corotines in the message callback but its not that easy either.

As I understand, you suggest to create a second engine for that usecase, right?

**Reply by aminalaee** [reply]
I think you don't need a second database engine, you can mix sync and async in your callbacks but that won't be very efficient.

Ideally you should replace the sync mqtt driver.

**Reply by nbraun-wolf** [reply]
How could I mix async and sync code in a sync callback? I cant use the await keyword in there. 

Yes Ideally I would replace the driver, I have looked into way for doing that as well, but there are not good options out there. Paho is very solid and battle proven. 

There async wrapper library but it tries to enforce the use of context managers, https://github.com/sbtinstruments/asyncio-mqtt. I didn't manage to really integrate this properly with fastapi. It's a long story.

**Reply by aminalaee** [reply]
something like this maybe:

```python
def callback():
    loop = asyncio.get_event_loop()
    loop.run_in_executor(None, async_func)
```

https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.run_in_executor

**Reply by nbraun-wolf** [reply]
I see, how about doing something like this?

```python
import asyncio
import json

from paho.mqtt.client import Client


def init_client(*topics):
    client = Client("test")
    client.connect_async("localhost")

    @client.connect_callback()
    def on_connect(client, userdata, flags, rc):
        print("Connection returned " + str(rc))
        client.subscribe(*topics)
    
    # get the main event loop (from fastapi)
    loop = asyncio.get_event_loop()

    def topic(sub):
        def decorator(coro):
            @client.topic_callback(sub)
            def handle(client, userdata, msg):
                # run the async callbacks on the main event loop
                asyncio.run_coroutine_threadsafe(coro(client, userdata, msg), loop)

        return decorator

    return client, topic

# async main program (potentially fastapi)
async def main():
    client, topic = init_client("my/topic")

    @topic("my/topic")
    async def print_payload(client, userdata, msg):
        await asyncio.sleep(2)
        print(msg.payload)

    client.loop_start()

    while True:
        print("running...")
        await asyncio.sleep(5)
        client.publish("my/topic", json.dumps({"mqtt": "msg"}))


asyncio.run(main())
```

**Reply by aminalaee** [reply]
I just took a quick look, I think it should be fine.

**Reply by nbraun-wolf** [reply]
Cool, thanks for taking your time and helping me to get on the right track.
